### ISSUE TITLE
Refactoring WorldLoader to Rhisis.Core

### ISSUE THREAD

**Issue by johmarjac** [Eastrall/Rhisis#52]
@Eastrall We already shared some minds yesterday about that. I ve created a new local branch refactor/resourceLoader where I am going to move the WorldLoader to Rhisis.Core and maybe rename it to ResourceLoader.

**Comment by Eastrall** [comment]
I am not sure that there is duplicate IDS in the define files. At the moment in the `WorldLoader`, there is a `Dictionnary<string, int>` for the defines. The key reprensents the name and the value the id.
In the `propItem.txt` you only have the IDS. 

When you load the items with the `ResourceTable` with the Defines and the Texts dictionaries, it will fetch automatically the good datas. So you don't have to do extra process to lookup into the dictionnaries yourself.

The `ResourceLoader` class in `Rhisis.Core` project would be nice. This will allow us to share the loading process among all tools.
But instead of naming it `ResourceLoader` maybe we could name it `GlobalResources` with a `Load()` method.

@Steve-Nzr what do you think about this ?

**Comment by johmarjac** [comment]
I just noticed it´s way more complicated to refactor it there than I thought as Maps cant be loaded anymore since maps requires the System and Systems are not in the Rhisis.Core. I am not sure if I got the time for that, I actually just want to move it because of redudancy when we need to load resources in another application besides World Server, in example Shop Editor. I think I will do some kind of provisoric solution for now locally for myself and later when Resourec Loading has been globalized I will apply changes to that and then make a PR of the Shop Editor.

**Comment by Eastrall** [comment]
It's true that the maps are loaded in the WorldLoader and the structure of the maps is in the `Rhisis.World` assembly.
We'll think about the maps later :wink:

**Comment by johmarjac** [comment]
I will for now copy WorldLoader to my ShopEditor project, remove the stuff that does not work and work on the Editor itself.

**Comment by Eastrall** [comment]
Yes. Use only what you need. 😄

**Comment by Eastrall** [comment]
I think it is a good moment to talk about this issue. I am working on the battle system (melee attack only) and I noticed that some formulas like `GetOriginHp`, `GetOriginMp` requires the job data stored in `dataSub1/propJob.inc`.
These formulas has to be used in the `ClusterServer` too when we create a new character to calculate the correct Hp, Mp and Fp the player needs at the begining of the adventure. By doing this, we could also imagine customizing the begin job at character creation; but that is not the discussion right now.

What I was thinking was introducing the concept of `Initializers` or `Loaders`.

We could have a class that inherits from `IInitializer` and implement a `Load` method. Then we can imagine a class that loads all initializers using reflection (only used at startup, so we don't really care about performances) and execute all initializers.

**Comment by Eastrall** [comment]
After some days of thinking, I've been able to find some solutions to make a refactoring to the game resources loader. If we think about it, each ressource can be loaded in different contexts: servers and tools. At the moment, the `ClusterServer` needs some resources already loaded in the `WorldServer`. This obviously ended in duplicating some code...
This is exactly what we want to avoid, because we now need to maintain two identical codes blocks.

Alright, there's my idea. Let's create a simple class named `GameResources` that would be a `Singleton` because resources need to be loaded only once and need to be accessible from everywhere.
Inside this `GameResources` class, we will find all the dictionaries that we have in the [WorldLoader](https://github.com/Eastrall/Rhisis/blob/develop/src/Rhisis.World/WorldLoader.cs#L53-L66) and more when we will add other resources.

This `GameResources` class will also have an `Load()` method that will process the loading of the resources. Thing is, we will not make a copy/paste from the `WorldLoader` to the `GameResources.Load()` method. We need to make the system wide open so we can load/initialize absolutly everything we want. To do so, let's introduce an interface named `IGameResourceLoader` which will contain only one method named `Load()`.

Then, the `GameResources` class will have an internal `List<IGameResourceLoader>` field that will contain every loaders, and in the `Load()` method, we will loop through this list and call the `Load()` method of each loader. 😉 
Of course, we will need to `Register()` the loader within the `GameResources` class... Let's resume !

GameResources class
```cs
class GameResources : Singleton
{
    private List<IGameResourceLoader> _loaders;

    public void Initialize() { ... }
    public void Register(IGameResourcerLoader loader) { ... }
}
```
IGameResourceLoader interface
```cs
interface IGameResourceLoader
{
    void Load();
}
```

And this is how we create loaders:
```cs
class DefineLoader : IGameResourceLoader
{
     public Load() { ... } // Load defines
}

class ItemsLoader : IGameResourceLoader
{
    public Load() { ... } // Load items
}
```

Any thoughts about this ? 😄

**Comment by johmarjac** [comment]
I think it's a good solution, but I am not yet sure about storing the dictionaries in the `GameResources `class for the following reasons:

1. Once the game resources are loaded, they stay in there forever.
2. `GameResources.Load()` will always load every resource loader which might not even be necessary.

Not sure if I am close enough with my brain, but what do you think about a `GameResources `class containing a generic `Load()` function returning the instance to a specific loader where as every loader itself contains his dictionary of resources it provides. We could also think of Lazy loading the resources first when necessary. But it's just a thought, I have not spent much time thinking about that.

**Comment by Eastrall** [comment]
I agree with you on the fact that the `GameResources` should not keep the list of the loaders. I've been thinking about an other approach that involves some dependency injection. We could imagine that all loaders can be registered in the `DependencyContainer` and then we could retrieve all of them by doing a simple `DependencyContainer.Instance.Resolve<IEnumerable<IGameResourceLoader>>()` or retrieve only one: `DependencyContainer.Instance.Resolve<DefineLoader>()`. This could fit what you were explaining in your comment above.

By doing this, we could also take advantage of the DI for injecting the `ILogger` interface to do some logging inside the loaders wihout referencing NLog in the `Rhisis.Core` assembly.

**Comment by johmarjac** [comment]
@Eastrall Yep, I would like that because in my opinion it behaves more modular depending the different resource models. However I do not know whether it fits your desire :)

**Comment by Eastrall** [comment]
I think this might fit the requirements. We can access the resources through the `DependencyContainer` and the `GameResources` properties.
Also, maybe we could use this to review the program Initialization process. I think the loading process should not be tied to the `WorldServer` class* it self. In my opinion, it should be outside. Maybe we could think about creating a fluent API like an ASP.NET Core program:
```cs
public static class Program
{
     public static void Main()
     {
          new AppBootstrapper()
              .SetConsoleTitle("...")
              .UseStartup<StartupClass>() // StartupClass: we could initialize dependency container, loaders, database configuration, ect...
              .Start<WorldServer>();
     }
}
```

The startup class will in fact initialize the dependency container, loaders and call the `GameResources.Load()` method and retrieve all loaders and load them. By doing this, we take full advantage of dependency injection. 😄

**Comment by johmarjac** [comment]
Yes, sounds and looks good, more clean and that is what we want.

**Comment by Eastrall** [comment]
So I came up with something today:
```cs
private static void Main()
{
    ConsoleAppBootstrapper.CreateApp()
        .SetConsoleTitle("Rhisis - World Server")
        .SetCulture("en-US")
        .UseStartup<WorldServerStartup>()
        .Run();
}
```
This code is going to be the entry point of every server. (Login, Cluster, World) It sets the console title, culture and usees a startup class that needs to implement the `IProgramStartup` interface described here:
```csharp
public interface IProgramStartup : IDisposable
{
    void Configure();
    void Run();
}
```
And this is how a typical server startup class will look like:
```cs
public sealed class ServerStartup : IProgramStartup
{
    public void Configure() { ... } // Initialize dependency container & load server/database configuration.
    public void Run() { ... } // Load game resources and start server using the loaded configuration
    public void Dispose() { ... }
}
```
Since we are using Dependency injection now, we can register everything we want in the `Configure()` method. Exemple:
```cs
public void Configure()
{
     DependencyContainer.Instance.Initialize();
     DependencyContainer.Instance.Register<IWorldServer, WorldServer>(ServiceLifetime.Singleton);

     foreach (var loaderType in this._loaders)
         DependencyContainer.Instance.Register(typeof(IGameResourceLoader), loaderType);
     ...
}

public void Run()
{
     var gameResources = DependencyContainer.Instance.Resolve<GameResources>();
     var server = DependencyContainer.Instance.Resolve<IWorldServer>();

    gameResources.Load();
    server.Start();
}
```
Since `GameResources` is registered in the `DependencyContainer` we can pass the loaders through constructor injection:
```cs
public class GameResources
{
    public GameResources(IEnumerable<IGameLoaderResources> loaders) { ... }
    public void Load() 
    { 
         loaders.ForEach(loader => loader.Load());
    }
}
```

Easy, right ? 😄

**Comment by johmarjac** [comment]
I think that's awesome! Good Job. I like that very much.

**Comment by Eastrall** [comment]
Solved in PR #134. Still a lot of loaders to make. Feel free to contribute.

Closing.
